### ISSUE TITLE
cqlsh doesn't work for `ImportError: No module named cqlshlib`

### ISSUE THREAD

**Issue by amoskong** [scylladb/scylla#4738]
*Installation details*
Scylla version (or git commit hash): 666.development-0.20190716.060e3f8ac
Cluster size: 1
OS (RHEL/CentOS/Ubuntu/AWS AMI): Ubuntu 18

The problem doesn't on `CentOS7`, cqlsh works well.

```
scylla-test@amos-ubuntu18-2:~$ sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 17723034C56D4B19
Executing: /tmp/apt-key-gpghome.OlRKJ11D64/gpg.1.sh --keyserver keyserver.ubuntu.com --recv-keys 17723034C56D4B19
gpg: key 17723034C56D4B19: public key "ScyllaDB Package Signing Key 2018 <security@scylladb.com>" imported
gpg: Total number processed: 1
gpg:               imported: 1

scylla-test@amos-ubuntu18-2:~$ sudo apt-get update

scylla-test@amos-ubuntu18-2:~$ sudo apt-get install scylla
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following package was automatically installed and is no longer required:
  grub-pc-bin
Use 'sudo apt autoremove' to remove it.
The following additional packages will be installed:
  ca-certificates-java fontconfig-config fonts-dejavu-core hwloc-nox java-common libavahi-client3 libavahi-common-data libavahi-common3 libcups2 libfontconfig1 libhwloc-plugins libhwloc5
  libjpeg-turbo8 libjpeg8 liblcms2-2 libltdl7 libpciaccess0 libpcsclite1 libpython-stdlib libpython2.7-minimal libpython2.7-stdlib libxi6 libxrender1 libxtst6 num-utils ocl-icd-libopencl1
  openjdk-8-jre-headless python python-minimal python-yaml python2.7 python2.7-minimal scylla-conf scylla-jmx scylla-kernel-conf scylla-python3 scylla-server scylla-tools scylla-tools-core
  x11-common
Suggested packages:
  default-jre cups-common libhwloc-contrib-plugins liblcms2-utils pcscd opencl-icd libnss-mdns fonts-dejavu-extra fonts-ipafont-gothic fonts-ipafont-mincho fonts-wqy-microhei
  fonts-wqy-zenhei fonts-indic python-doc python-tk python2.7-doc binutils binfmt-support
The following NEW packages will be installed:
  ca-certificates-java fontconfig-config fonts-dejavu-core hwloc-nox java-common libavahi-client3 libavahi-common-data libavahi-common3 libcups2 libfontconfig1 libhwloc-plugins libhwloc5
  libjpeg-turbo8 libjpeg8 liblcms2-2 libltdl7 libpciaccess0 libpcsclite1 libpython-stdlib libpython2.7-minimal libpython2.7-stdlib libxi6 libxrender1 libxtst6 num-utils ocl-icd-libopencl1
  openjdk-8-jre-headless python python-minimal python-yaml python2.7 python2.7-minimal scylla scylla-conf scylla-jmx scylla-kernel-conf scylla-python3 scylla-server scylla-tools
  scylla-tools-core x11-common
0 upgraded, 41 newly installed, 0 to remove and 32 not upgraded.
...
Setting up scylla-jmx (666.development-20190716.4303f06-1) ...
Setting up scylla-tools-core (666.development-20190716.bb68fe6-1) ...
Setting up scylla-tools (666.development-20190716.bb68fe6-1) ...
Setting up scylla (666.development-0.20190716.060e3f8ac-1) ...
Processing triggers for libc-bin (2.27-3ubuntu1) ...
Processing triggers for ureadahead (0.100.0-21) ...
Processing triggers for systemd (237-3ubuntu10.23) ...
Processing triggers for ca-certificates (20180409) ...
Updating certificates in /etc/ssl/certs...
0 added, 0 removed; done.
Running hooks in /etc/ca-certificates/update.d...

done.
done.
scylla-test@amos-ubuntu18-2:~$  cqlsh
Traceback (most recent call last):
  File "/usr/bin/cqlsh.py", line 169, in <module>
    from cqlshlib import cql3handling, cqlhandling, pylexotron, sslhandling, cqlshhandling
ImportError: No module named cqlshlib
scylla-test@amos-ubuntu18-2:~$ find /usr/lib/python2.7/site-packages/
/usr/lib/python2.7/site-packages/
/usr/lib/python2.7/site-packages/cassandra_pylib-0.0.0-py2.7.egg-info
/usr/lib/python2.7/site-packages/cqlshlib
/usr/lib/python2.7/site-packages/cqlshlib/saferscanner.py
/usr/lib/python2.7/site-packages/cqlshlib/copyutil.py
/usr/lib/python2.7/site-packages/cqlshlib/displaying.py
/usr/lib/python2.7/site-packages/cqlshlib/tracing.py
/usr/lib/python2.7/site-packages/cqlshlib/helptopics.py
/usr/lib/python2.7/site-packages/cqlshlib/sslhandling.py
/usr/lib/python2.7/site-packages/cqlshlib/wcwidth.py
/usr/lib/python2.7/site-packages/cqlshlib/cqlhandling.py
/usr/lib/python2.7/site-packages/cqlshlib/__init__.py
/usr/lib/python2.7/site-packages/cqlshlib/cql3handling.py
/usr/lib/python2.7/site-packages/cqlshlib/pylexotron.py
/usr/lib/python2.7/site-packages/cqlshlib/util.py
/usr/lib/python2.7/site-packages/cqlshlib/formatting.py
/usr/lib/python2.7/site-packages/cqlshlib/cqlshhandling.py

scylla-test@amos-ubuntu18-2:~$ dpkg-query -L scylla-tools|grep cqlsh
/usr/bin/cqlsh
/usr/bin/cqlsh.py
/usr/lib/python2.7/site-packages/cqlshlib
/usr/lib/python2.7/site-packages/cqlshlib/__init__.py
/usr/lib/python2.7/site-packages/cqlshlib/copyutil.py
/usr/lib/python2.7/site-packages/cqlshlib/cql3handling.py
/usr/lib/python2.7/site-packages/cqlshlib/cqlhandling.py
/usr/lib/python2.7/site-packages/cqlshlib/cqlshhandling.py
/usr/lib/python2.7/site-packages/cqlshlib/displaying.py
/usr/lib/python2.7/site-packages/cqlshlib/formatting.py
/usr/lib/python2.7/site-packages/cqlshlib/helptopics.py
/usr/lib/python2.7/site-packages/cqlshlib/pylexotron.py
/usr/lib/python2.7/site-packages/cqlshlib/saferscanner.py
/usr/lib/python2.7/site-packages/cqlshlib/sslhandling.py
/usr/lib/python2.7/site-packages/cqlshlib/tracing.py
/usr/lib/python2.7/site-packages/cqlshlib/util.py
/usr/lib/python2.7/site-packages/cqlshlib/wcwidth.py

scylla-test@amos-ubuntu18-2:/usr/lib/python2.7/site-packages$ ls -l cqlshlib/
total 288
-rw-r--r-- 1 root root 110269 Aug 24  2015 copyutil.py
-rw-r--r-- 1 root root  55422 Aug 24  2015 cql3handling.py
-rw-r--r-- 1 root root  13013 Aug 24  2015 cqlhandling.py
-rw-r--r-- 1 root root  10295 Aug 24  2015 cqlshhandling.py
-rw-r--r-- 1 root root   3977 Aug 24  2015 displaying.py
-rw-r--r-- 1 root root  23430 Aug 24  2015 formatting.py
-rw-r--r-- 1 root root   4532 Aug 24  2015 helptopics.py
-rw-r--r-- 1 root root    784 Aug 24  2015 __init__.py
-rw-r--r-- 1 root root  18855 Aug 24  2015 pylexotron.py
-rw-r--r-- 1 root root   2711 Aug 24  2015 saferscanner.py
-rw-r--r-- 1 root root   3727 Aug 24  2015 sslhandling.py
-rw-r--r-- 1 root root   3453 Aug 24  2015 tracing.py
-rw-r--r-- 1 root root   4525 Aug 24  2015 util.py
-rw-r--r-- 1 root root  15875 Aug 24  2015 wcwidth.py

scylla-test@amos-ubuntu18-2:/usr/lib/python2.7/site-packages$ whereis python2.7
python2: /usr/bin/python2.7 /usr/bin/python2 /usr/lib/python2.7 /etc/python2.7 /usr/local/lib/python2.7 /usr/share/man/man1/python2.1.gz

scylla-test@amos-ubuntu18-2:~$ dpkg -l |grep scylla
ii  scylla                         666.development-0.20190716.060e3f8ac-1      amd64        Scylla database metapackage
ii  scylla-conf                    666.development-0.20190716.060e3f8ac-1      amd64        Scylla database main configuration file
ii  scylla-jmx                     666.development-20190716.4303f06-1          all          Scylla JMX server binaries
ii  scylla-kernel-conf             666.development-0.20190716.060e3f8ac-1      amd64        Scylla kernel tuning configuration
ii  scylla-python3                 3.7.2-0.20190716.060e3f8ac-1                amd64        A standalone python3 interpreter that can be moved around different Linux machines
ii  scylla-server                  666.development-0.20190716.060e3f8ac-1      amd64        Scylla database server binaries
ii  scylla-tools                   666.development-20190716.bb68fe6-1          all          Scylla database tools
ii  scylla-tools-core              666.development-20190716.bb68fe6-1          all          Scylla database tools core files
```

@syuu1228

**Comment by slivne** [comment]
@amoskong just verifying - on 3.1 it works - right ?

**Comment by amoskong** [comment]
On Thu, Jul 18, 2019 at 3:04 PM Shlomi Livne <notifications@github.com>
wrote:

> @amoskong <https://github.com/amoskong> just verifying - on 3.1 it works
> - right ?
>
Yes, it works in 3.1

**Comment by amoskong** [comment]
I found this issue upgrade test of 3.1.0.rc4-0.20190828.d70c2db09

- Test ID: 660df9c0-489c-46ac-8441-069e9340f231
- http://jenkins.scylladb.com/view/scylla-3.1/job/scylla-3.1/view/scylla-3.1-test/job/rolling-upgrade/job/rolling-upgrade-ubuntu18.04/79/execution/node/40/log/
- https://cloudius-jenkins-test.s3.amazonaws.com/660df9c0-489c-46ac-8441-069e9340f231/job_log.zip
- full installation log: [scylla_issue_4738_ubuntu18_cqlshlib.txt](https://github.com/scylladb/scylla/files/3553867/scylla_issue_4738_ubuntu18_cqlshlib.txt)

```
scylla-test@amos-ubuntu18-tmp:~$ cqlsh
Traceback (most recent call last):
  File "/usr/bin/cqlsh.py", line 169, in <module>
    from cqlshlib import cql3handling, cqlhandling, pylexotron, sslhandling, cqlshhandling
ImportError: No module named cqlshlib

scylla-test@amos-ubuntu18-tmp:~$ dpkg -l |grep scylla
ii  scylla                         3.1.0.rc4-0.20190828.d70c2db09-1            amd64        Scylla database metapackage
ii  scylla-conf                    3.1.0.rc4-0.20190828.d70c2db09-1            amd64        Scylla database main configuration file
ii  scylla-jmx                     3.1.0~rc4-20190828.fb3d56d-1                all          Scylla JMX server binaries
ii  scylla-kernel-conf             3.1.0.rc4-0.20190828.d70c2db09-1            amd64        Scylla kernel tuning configuration
ii  scylla-python3                 3.7.2-0.20190828.d70c2db09-1                amd64        A standalone python3 interpreter that can be moved around different Linux machines
ii  scylla-server                  3.1.0.rc4-0.20190828.d70c2db09-1            amd64        Scylla database server binaries
ii  scylla-tools                   3.1.0~rc4-20190828.46280fbdaa-1             all          Scylla database tools
ii  scylla-tools-core              3.1.0~rc4-20190828.46280fbdaa-1             all          Scylla database tools core files

scylla-test@amos-ubuntu18-tmp:~$ python --version
Python 2.7.15+
scylla-test@amos-ubuntu18-tmp:~$ python
Python 2.7.15+ (default, Nov 27 2018, 23:36:35)
[GCC 7.3.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import cqlshlib
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: No module named cqlshlib
>>>

scylla-test@amos-ubuntu18-tmp:~$ find /usr/lib/|grep cqlshlib
/usr/lib/python2.7/site-packages/cqlshlib
/usr/lib/python2.7/site-packages/cqlshlib/helptopics.py
/usr/lib/python2.7/site-packages/cqlshlib/tracing.py
/usr/lib/python2.7/site-packages/cqlshlib/saferscanner.py
/usr/lib/python2.7/site-packages/cqlshlib/cqlshhandling.py
/usr/lib/python2.7/site-packages/cqlshlib/util.py
/usr/lib/python2.7/site-packages/cqlshlib/__init__.py
/usr/lib/python2.7/site-packages/cqlshlib/copyutil.py
/usr/lib/python2.7/site-packages/cqlshlib/pylexotron.py
/usr/lib/python2.7/site-packages/cqlshlib/wcwidth.py
/usr/lib/python2.7/site-packages/cqlshlib/formatting.py
/usr/lib/python2.7/site-packages/cqlshlib/cql3handling.py
/usr/lib/python2.7/site-packages/cqlshlib/displaying.py
/usr/lib/python2.7/site-packages/cqlshlib/cqlhandling.py
/usr/lib/python2.7/site-packages/cqlshlib/sslhandling.py
```

**Comment by amoskong** [comment]
@syuu1228 @slivne  We need fix this in 3.1

**Comment by glommer** [comment]
I have a patch (intended for post 3.1) that would probably fix it:
87dab77c1e797b13a8faebc19256e953e6b6b7e6

(That's issue #102)

It is intended for non-root relocatable but would likely work well in this
context too (since I also tested .deb and .rpm installations and it worked)

**Comment by slivne** [comment]
@avikivity do you want to take into 3.1 the patch from @glommer ?

**Comment by amoskong** [comment]
Related commit for scylla-tools-java:  https://github.com/scylladb/scylla-tools-java/commit/87dab77c1e797b13a8faebc19256e953e6b6b7e6

**Comment by slivne** [comment]
@avikivity please have a look

**Comment by avikivity** [comment]
Works on master.

**Comment by avikivity** [comment]
Well, with 16.04. Testing 18.04 now.

**Comment by avikivity** [comment]
Well, master works on 18.04 too.

**Comment by avikivity** [comment]
Reproduced in 3.1.

**Comment by avikivity** [comment]
master installs into /usr/lib/python2.7/dist-packages (which is in sys.path) while 3.1 installs into installs into /usr/lib/python2.7/site-packages (which isn't)

**Comment by avikivity** [comment]
scylladb/scylla-tools-java@2076d43196c00c0c6fea82793eb07ce05f1a6cfe does not help.

**Comment by avikivity** [comment]
Looks like debian patched python distutils to install in dist-packages instead of site-packages, and any attempt to cross-build (from Fedora) can't take that into account.

**Comment by avikivity** [comment]
> [scylladb/scylla-tools-java@2076d43](https://github.com/scylladb/scylla-tools-java/commit/2076d43196c00c0c6fea82793eb07ce05f1a6cfe) does not help.

It fixed one thing and broke another two.

**Comment by avikivity** [comment]
See scylladb/scylla-tools-java#112, scylladb/scylla-tools-java#113.

**Comment by amoskong** [comment]
> See [scylladb/scylla-tools-java#112](https://github.com/scylladb/scylla-tools-java/issues/112), [scylladb/scylla-tools-java#113](https://github.com/scylladb/scylla-tools-java/issues/113).

The two fixes are not included in 3.1.0.rc6, will retest when we have new packages.

**Comment by slivne** [comment]
patches are merged to 3.1 https://jenkins.scylladb.com/view/scylla-3.1/job/scylla-3.1/job/next/148/

we can test of jenkins - the fix will be part of 3.1.rc7

**Comment by amoskong** [comment]
The issue doesn't exist in `scylla-tools (3.1.0~rc7-20190910.32a2ff4d18-1)`. So it's fixed.
